File: connectors/integration-tests/src/test/scala/com/taskadapter/integrationtests/JiraRedmineIT.scala
package com.taskadapter.integrationtests

import com.taskadapter.connector.FieldRow
import com.taskadapter.connector.common.ProgressMonitorUtils
import com.taskadapter.connector.jira.{JiraConnector, JiraField}
import com.taskadapter.connector.redmine.{CustomFieldBuilder, RedmineConnector, RedmineField}
import com.taskadapter.connector.testlib.{TestSaver, TestUtils}
import com.taskadapter.core.TaskLoader
import com.taskadapter.model.{FieldRowBuilder, GTask, GUser}
import com.taskadapter.redmineapi.bean.{Issue, IssueFactory, Project}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll, FunSpec, Matchers}

import scala.collection.JavaConverters._

@RunWith(classOf[JUnitRunner])
class JiraRedmineIT extends FunSpec with Matchers with BeforeAndAfter with BeforeAndAfterAll {

  private var redmineProject: Option[Project] = None

  private val mgr = RedmineTestInitializer.mgr
  val redmineConfig = TestConfigs.getRedmineConfig
  val redmineConnector = new RedmineConnector(redmineConfig, TestConfigs.getRedmineServerInfo)
  val jiraConfig = TestConfigs.getJiraConfig
  val jiraSetup = TestConfigs.getJiraSetup
  val jiraConnector = new JiraConnector(jiraConfig, jiraSetup)
  val adapter = new Adapter(redmineConnector, jiraConnector)

  before {
    // have to create a project for each test, otherwise stuff created during one test interferes with others
    redmineProject = Some(RedmineTestInitializer.createProject)
    redmineConfig.setProjectKey(redmineProject.get.getIdentifier)
    redmineConfig.setFindUserByName(true)
  }

  after {
    RedmineTestInitializer.deleteProject(redmineProject.get.getIdentifier)
  }


  it("Description from Jira is saved to description in Redmine") {
    val rows = List(
      FieldRow(JiraField.summary, RedmineField.summary, ""),
      FieldRow(JiraField.description, RedmineField.description, "")
    )
    val task = new GTask()
    task.setValue(JiraField.summary, "summary1")
    task.setValue(JiraField.description, "description1")
    // create in Jira
    new TestSaver(redmineConnector, rows).saveAndLoad(task)

    // adapt
    val result = adapter.adapt(rows)

    // load from Redmine
    val redmine = TestUtils.loadCreatedTask(jiraConnector, rows.asJava, result)
    redmine.getValue(RedmineField.description) shouldBe "description1"
  }

  it("assignee and reporter can be loaded from JIRA and saved to Redmine") {
    val result = TestUtils.loadAndSave(jiraConnector, redmineConnector,
        Seq(RedmineField.summary, RedmineField.assignee, RedmineField.author)
    )
    val redmineAssignee = result.getValue(RedmineField.assignee).asInstanceOf[GUser]
    redmineAssignee.getDisplayName shouldBe "Redmine Admin"

    val redmineReporter = result.getValue(RedmineField.author).asInstanceOf[GUser]
    redmineReporter.getDisplayName shouldBe "Redmine Admin"
  }

  it("assignee can be loaded from Redmine and saved to JIRA") {
    val loadedTasks = TaskLoader.loadTasks(1, redmineConnector, "sourceName", ProgressMonitorUtils.DUMMY_MONITOR).asScala.toList

    val result = TestUtils.saveAndLoadList(jiraConnector, loadedTasks,
      FieldRowBuilder.rows(
        Seq(JiraField.summary, JiraField.assignee)
      )
    )
    val ass = result.head.getValue(JiraField.assignee).asInstanceOf[GUser]
    ass.getDisplayName shouldBe jiraSetup.userName

    val reporter = result.head.getValue(JiraField.reporter).asInstanceOf[GUser]
    reporter.getDisplayName shouldBe jiraSetup.userName
  }

  def createIssueInRedmineWithCustomField(fieldName: String, value: String): Issue = {
    val customFieldDefinitions = mgr.getCustomFieldManager.getCustomFieldDefinitions
    val issue = IssueFactory.create(redmineProject.get.getId, "some summary")
    CustomFieldBuilder.add(issue, customFieldDefinitions, fieldName, value)
    mgr.getIssueManager.createIssue(issue)
  }

  def createIssueInRedmine(description: String = ""): Issue = {
    val issue = IssueFactory.create(redmineProject.get.getId, "some summary")
    issue.setDescription(description)
    mgr.getIssueManager.createIssue(issue)
  }

}

File: connectors/integration-tests/src/test/java/com/taskadapter/core/JiraRedmineIT.java
package com.taskadapter.core;

import com.taskadapter.connector.FieldRow;
import com.taskadapter.connector.common.ProgressMonitorUtils;
import com.taskadapter.connector.definition.SaveResult;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.connector.definition.exceptions.ConnectorException;
import com.taskadapter.connector.jira.JiraConfig;
import com.taskadapter.connector.jira.JiraConnector;
import com.taskadapter.connector.redmine.RedmineConfig;
import com.taskadapter.connector.redmine.RedmineConnector;
import com.taskadapter.connector.testlib.TestSaver;
import com.taskadapter.connector.testlib.TestUtils;
import com.taskadapter.integrationtests.RedmineTestInitializer;
import com.taskadapter.integrationtests.TestConfigs;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import com.taskadapter.model.GTaskBuilder;
import com.taskadapter.redmineapi.RedmineException;
import com.taskadapter.redmineapi.bean.Issue;
import com.taskadapter.redmineapi.bean.Project;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static com.taskadapter.core.JavaFieldAdapter.AssigneeLoginNameOpt;
import static com.taskadapter.core.JavaFieldAdapter.ReporterLoginNameOpt;
import static com.taskadapter.core.JavaFieldAdapter.descriptionOpt;
import static com.taskadapter.core.JavaFieldAdapter.summaryOpt;
import static org.assertj.core.api.Assertions.assertThat;

public class JiraRedmineIT {

    private RedmineConfig sourceConfig = TestConfigs.getRedmineConfig();
    private RedmineConfig targetConfig = TestConfigs.getRedmineConfig();
    private RedmineConnector sourceRedmineConnector = new RedmineConnector(sourceConfig, TestConfigs.getRedmineSetup());
    private RedmineConnector targetRedmineConnector = new RedmineConnector(targetConfig, TestConfigs.getRedmineSetup());

    private Adapter adapter = new Adapter(sourceRedmineConnector, targetRedmineConnector);

    private RedmineConfig redmineConfigWithResolveAssignees = TestConfigs.getRedmineConfig();
    private RedmineConnector redmineConnectorWithResolveAssignees = new RedmineConnector(redmineConfigWithResolveAssignees, TestConfigs.getRedmineSetup());
    private Project redmineProject;

    private JiraConfig jiraConfig = TestConfigs.getJiraConfig();
    private WebConnectorSetup jiraSetup = TestConfigs.getJiraSetup();
    private JiraConnector jiraConnector = new JiraConnector(jiraConfig, jiraSetup);

    @Before
    public void beforeEachTest() {
        redmineConfigWithResolveAssignees.setFindUserByName(true);

        // have to create a project for each test, otherwise stuff created during one test interferes with others
        redmineProject = RedmineTestInitializer.createProject();
        sourceConfig.setProjectKey(redmineProject.getIdentifier());
        targetConfig.setProjectKey(redmineProject.getIdentifier());
        redmineConfigWithResolveAssignees.setProjectKey(redmineProject.getIdentifier());
    }

    @After
    public void afterEachTest() throws RedmineException {
        RedmineTestInitializer.deleteProject(redmineProject.getIdentifier());
    }

    // Description from Jira is saved to description in Redmine
    @Test
    public void descriptionFromJiraIsSavedToRedmine() throws ConnectorException {
        List<FieldRow<?>> rows = Arrays.asList(
                new FieldRow(summaryOpt, summaryOpt, ""),
                new FieldRow(descriptionOpt, descriptionOpt, "")
        );
        GTask task = new GTask();
        task.setValue(AllFields.summary, "summary1");
        task.setValue(AllFields.description, "description1");
        new TestSaver(redmineConnectorWithResolveAssignees, rows).saveAndLoad(task);

        SaveResult result = adapter.adapt(rows);

        GTask redmine = TestUtils.loadCreatedTask(redmineConnectorWithResolveAssignees, rows, result);
        assertThat(redmine.getValue(AllFields.description)).isEqualTo("description1");
    }

    // assignee and reporter can be loaded from JIRA and saved to Redmine
    @Test
    public void assigneeAndReporterCanBeLoadedFromJiraAndSavedToRedmine() throws ConnectorException {
        List<FieldRow<?>> rows = Arrays.asList(new FieldRow(summaryOpt, summaryOpt, ""),
                new FieldRow(AssigneeLoginNameOpt, AssigneeLoginNameOpt, null),
                new FieldRow(ReporterLoginNameOpt, ReporterLoginNameOpt, null)
        );

        GTask fromJira = TestUtils.saveAndLoad(jiraConnector,
                new GTaskBuilder()
                        .withRandom(AllFields.summary)
                        .withAssigneeLogin(RedmineTestInitializer.currentUser.getLoginName())
                        .build(),
                rows);

        GTask redmineResult = TestUtils.saveAndLoad(redmineConnectorWithResolveAssignees, fromJira, rows);

        assertThat(redmineResult.getValue(AllFields.assigneeFullName)).isEqualTo("Redmine Admin");
        assertThat(redmineResult.getValue(AllFields.assigneeLoginName)).isEqualTo("user");

        assertThat(redmineResult.getValue(AllFields.reporterFullName)).isEqualTo("Redmine Admin");
        assertThat(redmineResult.getValue(AllFields.reporterLoginName)).isEqualTo("user");
    }

    // assignee can be loaded from Redmine and saved to JIRA
    @Test
    public void assigneeCanBeLoadedFromRedmineAndSavedToJira() throws Exception {
        Issue created = RedmineTestUtil.createIssueInRedmine(redmineProject, "some description",
                RedmineTestInitializer.currentUser);
        List<GTask> loadedTasks = TaskLoader.loadTasks(1, redmineConnectorWithResolveAssignees, "sourceName", ProgressMonitorUtils.DUMMY_MONITOR);
        assertThat(loadedTasks).hasSize(1);
        GTask redmineTask = loadedTasks.get(0);
        assertThat(redmineTask.getValue(AllFields.assigneeLoginName))
                .isEqualTo(RedmineTestInitializer.currentUser.getLoginName());

        GTask result = TestUtils.saveAndLoad(jiraConnector, redmineTask,
                Arrays.asList(
                        new FieldRow(summaryOpt, summaryOpt, ""),
                        new FieldRow(AssigneeLoginNameOpt, AssigneeLoginNameOpt, null)
                )
        );
        assertThat(result.getValue(AllFields.assigneeLoginName)).isEqualTo(jiraSetup.getUserName());
        assertThat(result.getValue(AllFields.reporterFullName)).isEqualTo(jiraSetup.getUserName());
    }
}
